File: Midterm-Finals Review/LL.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        
class LL:
    def __init__(self):
        self.head = None
    
    def insert(self, idx, data):
        new_node = Node(data)
        
        if self.head == None:
            self.head = new_node
            
        else:
            if idx == 0:
                head_temp = self.head
                self.head = new_node
                self.head.next = head_temp
            else:
                cur = self.head
                for i in range(idx - 1):
                    cur = cur.next

                cur.next, new_node.next = new_node, cur.next
    
    def is_in(self, data):
        cur = self.head
        while cur:
            if cur.data == data:
                return True
            cur = cur.next
        return False
    
    def remove(self, data):
        if (self.head.data == data):
            self.head = self.head.next
            return
        
        cur = self.head
        while cur.next != None and cur.next.data != data:
            cur = cur.next
        
        if cur.next == None:
            print("Data not found")
            return
        cur.next = (cur.next).next
            
        
        
    def print(self):
        cur = self.head
        while(cur != None):
            print (cur.data, "->", end = "")
            cur = cur.next
        print("None")
            

if __name__ == '__main__':
    myLL = LL()
    myLL.insert(0,123)
    myLL.insert(0,125)
    myLL.print()

File: Midterm-Finals Review/LL.c
#include <stdio.h>
#include <stdlib.h>

// Node structure
struct Node {
    int data;
    struct Node* next;
};

// Function to create a new node
struct Node* create_node(int data) {
    struct Node* new_node = (struct Node*) malloc(sizeof(struct Node));
    new_node->data = data;
    new_node->next = NULL;
    return new_node;
}

// Append node and update size
void append_node(struct Node** head_ref, int data, int* size) {
    struct Node* new_node = create_node(data);

    if (*head_ref == NULL) {
        *head_ref = new_node;
    } else {
        struct Node* temp = *head_ref;
        while (temp->next != NULL) {
            temp = temp->next;
        }
        temp->next = new_node;
    }

    (*size)++; // increment size
}

// Print list
void print_list(struct Node* head) {
    struct Node* current = head;
    while (current != NULL) {
        printf("%d -> ", current->data);
        current = current->next;
    }
    printf("NULL\n");
}

int hasLoop(struct Node* head){
    struct Node* fast = head;
    struct Node* slow = head;

    while (fast == NULL && fast -> next == NULL){
        if (fast == slow){
            return 1;
        }
            fast = (fast -> next)-> next;
            slow = slow -> next;
    }
    return NULL; 
}

void loopLocation(struct Node* head){
    struct Node* cur = head;
    
    if (hasLoop(head) == 1){
        while (cur != NULL)
    }
    return;
}

void removeLoop(struct Node* head, struct node* ){
    struct Node* current = head;
    while current
}